build-system/common/transform-cache.js

```diff
@@ -27,35 +29,45 @@ class TransformCache {
 
   /**
    * @param {string} hash
-   * @return {null|Promise<Buffer|string>}
+   * @return {null|Promise<T>}
    */
   get(hash) {
     const cached = this.transformMap.get(hash);
     if (cached) {
       return cached;
     }
-    const filename = hash + this.fileExtension;
+
+    const filename = this.key_(hash);
     if (this.fsCache.has(filename)) {
-      const transformedPromise = fs.readFile(
-        path.join(this.cacheDir, filename)
-      );
-      this.transformMap.set(hash, transformedPromise);
-      return transformedPromise;
+      const persisted = fs.readJson(path.join(this.cacheDir, filename));
+      this.transformMap.set(hash, persisted);
+      return persisted;
     }
+
     return null;
   }
 
   /**
    * @param {string} hash
-   * @param {Promise<string>} transformPromise
+   * @param {Promise<T>} transformPromise
    */
   set(hash, transformPromise) {
     if (this.transformMap.has(hash)) {
       throw new Error(`Read race: Attempting to transform ${hash} file twice.`);
     }
     this.transformMap.set(hash, transformPromise);
-    const filepath = path.join(this.cacheDir, hash) + this.fileExtension;
-    transformPromise.then((contents) => fs.outputFile(filepath, contents));
+
+    const filepath = path.join(this.cacheDir, this.key_(hash));
+    transformPromise.then((contents) => fs.outputJson(filepath, contents));
```

Comment:
This changes `TransformCache` to use JSON so that we can persist both the transformed code and the sourcemap in the cache directory.